pkg/actors/targets/workflow/activity.go

```diff
@@ -0,0 +1,485 @@
+
+	scheduler          todo.ActivityScheduler
+	state              *activityState
+	cachingDisabled    bool
+	defaultTimeout     time.Duration
+	reminderInterval   time.Duration
+	completed          atomic.Bool
+	schedulerReminders bool
+}
+
+// ActivityRequest represents a request by a worklow to invoke an activity.
+type ActivityRequest struct {
+	HistoryEvent []byte
+}
+
+type activityState struct {
+	EventPayload []byte
+}
+
+type ActivityOptions struct {
+	AppID              string
+	ActivityActorType  string
+	WorkflowActorType  string
+	CachingDisabled    bool
+	DefaultTimeout     *time.Duration
+	ReminderInterval   *time.Duration
+	Scheduler          todo.ActivityScheduler
+	Actors             actors.Interface
+	SchedulerReminders bool
+}
+
+func ActivityFactory(opts ActivityOptions) targets.Factory {
+	return func(actorID string) targets.Interface {
+		reminderInterval := time.Minute * 1
+		defaultTimeout := time.Hour * 1
+
+		if opts.ReminderInterval != nil {
+			reminderInterval = *opts.ReminderInterval
+		}
+		if opts.DefaultTimeout != nil {
+			defaultTimeout = *opts.DefaultTimeout
+		}
+
+		return &activity{
+			appID:              opts.AppID,
+			actorID:            actorID,
+			actorType:          opts.ActivityActorType,
+			cachingDisabled:    opts.CachingDisabled,
+			workflowActorType:  opts.WorkflowActorType,
+			reminderInterval:   reminderInterval,
+			defaultTimeout:     defaultTimeout,
+			actors:             opts.Actors,
+			scheduler:          opts.Scheduler,
+			schedulerReminders: opts.SchedulerReminders,
+			lock:               internal.NewLock(internal.LockOptions{ActorType: opts.ActivityActorType}),
+		}
+	}
+}
+
+// InvokeMethod implements actors.InternalActor and schedules the background execution of a workflow activity.
+// Activities are scheduled by workflows and can execute for arbitrary lengths of time. Instead of executing
+// activity logic directly, InvokeMethod creates a reminder that executes the activity logic. InvokeMethod
+// returns immediately after creating the reminder, enabling the workflow to continue processing other events
+// in parallel.
+func (a *activity) InvokeMethod(ctx context.Context, req *internalsv1pb.InternalInvokeRequest) (*internalsv1pb.InternalInvokeResponse, error) {
+	log.Debugf("Activity actor '%s': invoking method '%s'", a.actorID, req.GetMessage().GetMethod())
+
+	imReq, err := invokev1.FromInternalInvokeRequest(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create InvokeMethodRequest: %w", err)
+	}
+	defer imReq.Close()
+
+	cancel, err := a.lock.LockRequest(imReq)
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+
+	msg := imReq.Message()
+
+	var ar ActivityRequest
+	if err = gob.NewDecoder(bytes.NewReader(msg.GetData().GetValue())).Decode(&ar); err != nil {
+		return nil, fmt.Errorf("failed to decode activity request: %w", err)
+	}
+
+	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
```

Comment:
Based on this comment, should we do a check on the returned state value from the `a.loadActivityState` call?

Reply:
This will be removed, state is moving into reminders.